File: windows/codex_session_manager_electron/src/update/release-manifest.js
const crypto = require('node:crypto');

const INVALID_MESSAGE = '更新信息验证失败，请联系管理员';
const SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');
const VERSION = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/;
const SHA256 = /^[a-f0-9]{64}$/;
const PUBLISHED_AT = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{3})?Z$/;
const TOP_FIELDS = [
  'schemaVersion', 'channel', 'version', 'build',
  'publishedAt', 'required', 'notes', 'platforms',
];
const ARTIFACT_FIELDS = ['url', 'size', 'sha256'];
const PLATFORM_KEYS = ['macos-arm64', 'windows-x64'];

function invalid(reason) {
  return new Error(`Invalid release manifest: ${reason}`);
}

function exactKeys(value, expected, scope) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw invalid(`${scope} must be an object`);
  }
  const actual = Object.keys(value).sort();
  const wanted = [...expected].sort();
  if (actual.length !== wanted.length || actual.some((key, index) => key !== wanted[index])) {
    const unknown = actual.find((key) => !wanted.includes(key));
    if (unknown) throw invalid(`unknown ${scope} field ${unknown}`);
    throw invalid(`missing ${scope} fields`);
  }
}

function parseVersion(value) {
  if (typeof value !== 'string' || !VERSION.test(value)) {
    throw invalid('version must use numeric X.Y.Z format');
  }
  const components = value.split('.').map(Number);
  if (components.some((component) => !Number.isSafeInteger(component))) {
    throw invalid('version components must be safe integers');
  }
  return components;
}

function validatePublishedAt(value) {
  if (typeof value !== 'string' || !PUBLISHED_AT.test(value)) {
    throw invalid('publishedAt must be an ISO-8601 UTC timestamp');
  }
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) {
    throw invalid('publishedAt must be an ISO-8601 UTC timestamp');
  }
  const canonical = value.includes('.')
    ? date.toISOString()
    : date.toISOString().replace('.000Z', 'Z');
  if (canonical !== value) {
    throw invalid('publishedAt must be an ISO-8601 UTC timestamp');
  }
}

function validateRelativeArtifactURL(value) {
  if (typeof value !== 'string'
      || value.length === 0
      || value.startsWith('/')
      || value.includes('\\')
      || value.includes('?')
      || value.includes('#')) {
    throw invalid('artifact url must be a safe relative URL');
  }

  let parsed;
  try {
    parsed = new URL(value, 'https://release.invalid/');
  } catch {
    throw invalid('artifact url must be a safe relative URL');
  }
  if (parsed.origin !== 'https://release.invalid' || parsed.username || parsed.password) {
    throw invalid('artifact url must be a safe relative URL');
  }

  const segments = value.split('/');
  if (segments.some((segment) => segment.length === 0)) {
    throw invalid('artifact url must be a safe relative URL');
  }
  for (const segment of segments) {
    let decoded;
    try {
      decoded = decodeURIComponent(segment);
    } catch {
      throw invalid('artifact url must be a safe relative URL');
    }
    if (decoded === '.' || decoded === '..' || decoded.includes('/') || decoded.includes('\\')) {
      throw invalid('artifact url must be a safe relative URL');
    }
  }
}

function validateArtifact(artifact) {
  exactKeys(artifact, ARTIFACT_FIELDS, 'artifact');
  validateRelativeArtifactURL(artifact.url);
  if (!Number.isSafeInteger(artifact.size) || artifact.size <= 0) {
    throw invalid('artifact size must be a positive safe integer');
  }
  if (typeof artifact.sha256 !== 'string' || !SHA256.test(artifact.sha256)) {
    throw invalid('artifact sha256 must be 64 lowercase hexadecimal characters');
  }
}

function validateManifest(manifest) {
  exactKeys(manifest, TOP_FIELDS, 'top level');
  if (manifest.schemaVersion !== 1) throw invalid('schemaVersion must be 1');
  if (manifest.channel !== 'stable') throw invalid('channel must be stable');
  parseVersion(manifest.version);
  if (!Number.isSafeInteger(manifest.build) || manifest.build <= 0) {
    throw invalid('build must be a positive safe integer');
  }
  validatePublishedAt(manifest.publishedAt);
  if (manifest.required !== false) throw invalid('required must be false');
  if (!Array.isArray(manifest.notes) || manifest.notes.length > 20) {
    throw invalid('notes must contain at most 20 entries');
  }
  if (manifest.notes.some((note) => typeof note !== 'string'
      || Buffer.byteLength(note, 'utf8') > 500)) {
    throw invalid('every note must be at most 500 UTF-8 bytes');
  }
  exactKeys(manifest.platforms, PLATFORM_KEYS, 'platforms');
  for (const platform of PLATFORM_KEYS) validateArtifact(manifest.platforms[platform]);
  return manifest;
}

function decodeCanonicalBase64(value, expectedLength, label) {
  const encoded = Buffer.isBuffer(value) ? value.toString('utf8').trim() : String(value).trim();
  const decoded = Buffer.from(encoded, 'base64');
  if (decoded.length !== expectedLength || decoded.toString('base64') !== encoded) {
    throw invalid(`${label} is malformed`);
  }
  return decoded;
}

function parseAndVerifyManifest(manifestBytes, signatureBytes, publicKeyBase64) {
  const bytes = Buffer.from(manifestBytes);
  const signature = decodeCanonicalBase64(signatureBytes, 64, 'signature');
  const rawPublicKey = decodeCanonicalBase64(publicKeyBase64, 32, 'public key');
  const publicKey = crypto.createPublicKey({
    key: Buffer.concat([SPKI_PREFIX, rawPublicKey]),
    format: 'der',
    type: 'spki',
  });
  if (!crypto.verify(null, bytes, publicKey, signature)) {
    throw invalid('signature verification failed');
  }

  let manifest;
  try {
    manifest = JSON.parse(bytes.toString('utf8'));
  } catch {
    throw invalid('JSON does not match schema version 1');
  }
  return validateManifest(manifest);
}

function compareVersions(lhs, rhs) {
  const left = parseVersion(lhs);
  const right = parseVersion(rhs);
  for (let index = 0; index < 3; index += 1) {
    if (left[index] !== right[index]) return left[index] < right[index] ? -1 : 1;
  }
  return 0;
}

function selectUpdate(manifest, currentVersion, currentBuild, platform) {
  try {
    validateManifest(manifest);
    parseVersion(currentVersion);
    if (!Number.isSafeInteger(currentBuild) || currentBuild < 0) {
      throw invalid('current build must be a non-negative safe integer');
    }
    if (!PLATFORM_KEYS.includes(platform)) throw invalid('unsupported platform');

    const comparison = compareVersions(manifest.version, currentVersion);
    if (comparison > 0 || (comparison === 0 && manifest.build > currentBuild)) {
      return { status: 'available', manifest, artifact: manifest.platforms[platform] };
    }
    if (comparison === 0 && manifest.build === currentBuild) {
      return { status: 'up-to-date' };
    }
    return { status: 'invalid', message: INVALID_MESSAGE };
  } catch {
    return { status: 'invalid', message: INVALID_MESSAGE };
  }
}

module.exports = {
  INVALID_MESSAGE,
  parseAndVerifyManifest,
  selectUpdate,
};

File: Sources/CodexSessionVaultCore/Update/ReleaseManifest.swift
import Foundation

public enum ReleaseManifestError: Error, Equatable, LocalizedError, Sendable {
    case invalid(String)

    public var errorDescription: String? {
        switch self {
        case .invalid(let reason):
            return "Invalid release manifest: \(reason)"
        }
    }
}

public enum UpdatePlatform: String, Codable, Sendable {
    case macosArm64 = "macos-arm64"
    case windowsX64 = "windows-x64"
}

public struct SemanticVersion: Comparable, Equatable, Sendable {
    public let major: Int
    public let minor: Int
    public let patch: Int

    public init(_ value: String) throws {
        let components = value.split(separator: ".", omittingEmptySubsequences: false)
        guard components.count == 3 else {
            throw ReleaseManifestError.invalid("version must use numeric X.Y.Z format")
        }
        let parsed = try components.map { component -> Int in
            guard !component.isEmpty,
                  component.allSatisfy(\.isNumber),
                  component == "0" || component.first != "0",
                  let number = Int(component) else {
                throw ReleaseManifestError.invalid("version must use numeric X.Y.Z format")
            }
            return number
        }
        self.major = parsed[0]
        self.minor = parsed[1]
        self.patch = parsed[2]
    }

    public static func < (lhs: SemanticVersion, rhs: SemanticVersion) -> Bool {
        if lhs.major != rhs.major { return lhs.major < rhs.major }
        if lhs.minor != rhs.minor { return lhs.minor < rhs.minor }
        return lhs.patch < rhs.patch
    }
}

public struct ReleaseArtifact: Codable, Equatable, Sendable {
    public let url: String
    public let size: Int64
    public let sha256: String

    public init(url: String, size: Int64, sha256: String) throws {
        self.url = url
        self.size = size
        self.sha256 = sha256
        try validate()
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DynamicCodingKey.self)
        try requireExactKeys(container.allKeys, expected: ["url", "size", "sha256"], scope: "artifact")
        self.url = try container.decode(String.self, forKey: DynamicCodingKey("url"))
        self.size = try container.decode(Int64.self, forKey: DynamicCodingKey("size"))
        self.sha256 = try container.decode(String.self, forKey: DynamicCodingKey("sha256"))
        try validate()
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: DynamicCodingKey.self)
        try container.encode(url, forKey: DynamicCodingKey("url"))
        try container.encode(size, forKey: DynamicCodingKey("size"))
        try container.encode(sha256, forKey: DynamicCodingKey("sha256"))
    }

    fileprivate func validate() throws {
        guard size > 0 else {
            throw ReleaseManifestError.invalid("artifact size must be positive")
        }
        guard sha256.count == 64,
              sha256.unicodeScalars.allSatisfy({ scalar in
                  ("0"..."9").contains(Character(String(scalar)))
                      || ("a"..."f").contains(Character(String(scalar)))
              }) else {
            throw ReleaseManifestError.invalid("artifact sha256 must be 64 lowercase hexadecimal characters")
        }
        try validateRelativeArtifactURL(url)
    }
}

public struct ReleaseManifest: Codable, Equatable, Sendable {
    public let schemaVersion: Int
    public let channel: String
    public let version: String
    public let build: Int
    public let publishedAt: Date
    public let required: Bool
    public let notes: [String]
    public let platforms: [String: ReleaseArtifact]

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DynamicCodingKey.self)
        try requireExactKeys(
            container.allKeys,
            expected: [
                "schemaVersion", "channel", "version", "build",
                "publishedAt", "required", "notes", "platforms"
            ],
            scope: "top level"
        )
        self.schemaVersion = try container.decode(Int.self, forKey: DynamicCodingKey("schemaVersion"))
        self.channel = try container.decode(String.self, forKey: DynamicCodingKey("channel"))
        self.version = try container.decode(String.self, forKey: DynamicCodingKey("version"))
        self.build = try container.decode(Int.self, forKey: DynamicCodingKey("build"))
        let publishedAtString = try container.decode(String.self, forKey: DynamicCodingKey("publishedAt"))
        self.publishedAt = try parsePublishedAt(publishedAtString)
        self.required = try container.decode(Bool.self, forKey: DynamicCodingKey("required"))
        self.notes = try container.decode([String].self, forKey: DynamicCodingKey("notes"))
        self.platforms = try container.decode(
            [String: ReleaseArtifact].self,
            forKey: DynamicCodingKey("platforms")
        )
        try validate()
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: DynamicCodingKey.self)
        try container.encode(schemaVersion, forKey: DynamicCodingKey("schemaVersion"))
        try container.encode(channel, forKey: DynamicCodingKey("channel"))
        try container.encode(version, forKey: DynamicCodingKey("version"))
        try container.encode(build, forKey: DynamicCodingKey("build"))
        try container.encode(formatPublishedAt(publishedAt), forKey: DynamicCodingKey("publishedAt"))
        try container.encode(required, forKey: DynamicCodingKey("required"))
        try container.encode(notes, forKey: DynamicCodingKey("notes"))
        try container.encode(platforms, forKey: DynamicCodingKey("platforms"))
    }

    public static func decodeValidated(from data: Data) throws -> ReleaseManifest {
        do {
            return try JSONDecoder().decode(ReleaseManifest.self, from: data)
        } catch let error as ReleaseManifestError {
            throw error
        } catch {
            throw ReleaseManifestError.invalid("JSON does not match schema version 1")
        }
    }

    public func validated() throws -> ReleaseManifest {
        try validate()
        return self
    }

    public func artifact(for platform: UpdatePlatform) throws -> ReleaseArtifact {
        guard let artifact = platforms[platform.rawValue] else {
            throw ReleaseManifestError.invalid("missing artifact for \(platform.rawValue)")
        }
        return artifact
    }

    public func matchesSparkleItem(displayVersion: String, buildVersion: String) -> Bool {
        displayVersion == version && buildVersion == String(build)
    }

    private func validate() throws {
        guard schemaVersion == 1 else {
            throw ReleaseManifestError.invalid("schemaVersion must be 1")
        }
        guard channel == "stable" else {
            throw ReleaseManifestError.invalid("channel must be stable")
        }
        _ = try SemanticVersion(version)
        guard build > 0 else {
            throw ReleaseManifestError.invalid("build must be positive")
        }
        guard required == false else {
            throw ReleaseManifestError.invalid("required must be false")
        }
        guard notes.count <= 20 else {
            throw ReleaseManifestError.invalid("notes must contain at most 20 entries")
        }
        guard notes.allSatisfy({ $0.lengthOfBytes(using: .utf8) <= 500 }) else {
            throw ReleaseManifestError.invalid("every note must be at most 500 UTF-8 bytes")
        }
        let expectedPlatforms = Set([UpdatePlatform.macosArm64.rawValue, UpdatePlatform.windowsX64.rawValue])
        guard Set(platforms.keys) == expectedPlatforms else {
            throw ReleaseManifestError.invalid("platforms must contain macos-arm64 and windows-x64 only")
        }
        try platforms.values.forEach { try $0.validate() }
    }
}

private struct DynamicCodingKey: CodingKey, Hashable {
    let stringValue: String
    let intValue: Int? = nil

    init(_ stringValue: String) {
        self.stringValue = stringValue
    }

    init?(stringValue: String) {
        self.init(stringValue)
    }

    init?(intValue: Int) {
        return nil
    }
}

private func requireExactKeys(
    _ keys: [DynamicCodingKey],
    expected: Set<String>,
    scope: String
) throws {
    let actual = Set(keys.map(\.stringValue))
    guard actual == expected else {
        let unknown = actual.subtracting(expected).sorted()
        let missing = expected.subtracting(actual).sorted()
        if let field = unknown.first {
            throw ReleaseManifestError.invalid("unknown \(scope) field \(field)")
        }
        throw ReleaseManifestError.invalid("missing \(scope) fields \(missing.joined(separator: ", "))")
    }
}

private func validateRelativeArtifactURL(_ value: String) throws {
    guard !value.isEmpty,
          !value.hasPrefix("/"),
          !value.contains("\\"),
          !value.contains("?"),
          !value.contains("#"),
          URL(string: value)?.scheme == nil else {
        throw ReleaseManifestError.invalid("artifact url must be a safe relative URL")
    }
    let segments = value.split(separator: "/", omittingEmptySubsequences: false)
    guard !segments.contains(where: { $0.isEmpty }) else {
        throw ReleaseManifestError.invalid("artifact url must be a safe relative URL")
    }
    for segment in segments {
        guard let decoded = String(segment).removingPercentEncoding,
              decoded != ".",
              decoded != "..",
              !decoded.contains("/"),
              !decoded.contains("\\") else {
            throw ReleaseManifestError.invalid("artifact url must be a safe relative URL")
        }
    }
}

private func parsePublishedAt(_ value: String) throws -> Date {
    let pattern = #"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{3})?Z$"#
    guard value.range(of: pattern, options: .regularExpression) != nil else {
        throw ReleaseManifestError.invalid("publishedAt must be an ISO-8601 UTC timestamp")
    }
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = value.contains(".")
        ? [.withInternetDateTime, .withFractionalSeconds]
        : [.withInternetDateTime]
    guard let date = formatter.date(from: value) else {
        throw ReleaseManifestError.invalid("publishedAt must be an ISO-8601 UTC timestamp")
    }
    return date
}

private func formatPublishedAt(_ value: Date) -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime]
    return formatter.string(from: value)
}
